### ISSUE TITLE
sigv4 remote write does not deduplicate slashes

### ISSUE THREAD

**Issue by nfi-opsguru** [prometheus/prometheus#10200]
**What did you do?**

Configured Remote Write to Amazon Managed Prometheus using sigv4 creds presented by EKS.

**What did you expect to see?**

Remote writes to go through.

**What did you see instead? Under which circumstances?**

Remote write failed 100% of the time with a sigv4 signature failure. I noticed that the URL had two slashes in it (due to templating). Once I fixed this, it worked. 

According to the sigv4 docs, this URI should have the slashes deduplicated.

**Environment**

Kubernetes (EKS) running linux amd64

* System information:

```
uname -srm
Linux 5.4.162-86.275.amzn2.x86_64 x86_64
```

* Prometheus version:

Broken in build from `main`:
```
# git describe --tag
v2.33.0-rc.1-10-ga4d320661
```

And at least
`quay.io/prometheus/prometheus:v2.32.1@sha256:82252381f9645bfdb0d0cee6dbe05505d86b9a784d4d0004e8b731705fb0ce11`

* Prometheus configuration file:
```
...
remote_write:
- sigv4:
    region: us-east-2
    url: https://aps-workspaces.us-east-2.amazonaws.com/workspaces/<REDACTED WORKSPACE ID>//api/v1/remote_write
...
```

Notice the double slash in the path.

* Logs:
```
ts=2022-01-25T00:11:44.163Z caller=dedupe.go:112 component=remote level=error remote_name=ddbb0e url=https://aps-workspaces.us-east-2.amazonaws.com/workspaces/<REDACTED WORKSPACE ID>//api/v1/remote_write msg="non-recoverable error while sending metadata" count=204 err="server returned HTTP status 403 Forbidden: {\"message\":\"The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.\\n\\nThe Canonical String for this request should have been\\n'POST\\n/workspaces/<REDACTED WORKSPACE ID/api/v1/remote_write\\n\\ncontent-encoding:snappy\\ncontent-type:application/x-protobuf\\nhost:aps-workspaces.us-east-2.amazonaws.com\\nx-amz-date:20220125T001144Z\\nx-amz-security-token:<REDACTED>"
```

* Hacks:

I patched `common/sigv4/sigv4.go` to enable extra AWS SDK debugging:

```
	signer := signer.NewSigner(signerCreds)
	signer.Logger = aws.NewDefaultLogger()
	signer.Debug = aws.LogDebugWithSigning
```

** Guesses **

Could be a bug in the SDK? I had a quick look and couldn't find any test cases specifically for duplicated slashes. https://github.com/aws/aws-sdk-go/tree/main/aws/signer/v4

**Comment by nfi-opsguru** [comment]
The [docs](https://docs.aws.amazon.com/sdk-for-go/api/aws/signer/v4/#pkg-overview) seem to suggest some amount of escaping is up to the caller, but prometheus probably wouldn't be able to determine how much escaping to do generally :/ So I guess it is kind of up to the user? Maybe just a warning in the docs would suffice?

**Comment by roidelapluie** [comment]
According to the documentation, we should call http.EscapePath ourselves when doing the sigv4 sign.

I am moving this issue to prometheus/common as that is where the code lives.

**Comment by importhuman** [comment]
Hello! Can I pick this up? @roidelapluie

**Comment by roidelapluie** [comment]
Yes! The code is at https://github.com/prometheus/common/tree/main/sigv4

**Comment by importhuman** [comment]
Don't we want to remove the extra double slash instead of escaping it? 
From the [AWS doc](https://docs.aws.amazon.com/sdk-for-go/api/aws/signer/v4/#pkg-overview), wouldn't `"//<hostname>/<path>"` correspond to `//aps-workspaces.us-east-2.amazonaws.com/workspaces/...`? 

Or do we want to just escape the extra backslash in `<REDACTED WORKSPACE ID>//api/...`? as per this line from the docs:

> AWS v4 signature validation requires that the canonical string's URI path element must be the URI escaped form of the HTTP request's path.

Is this backslash coming from AWS itself and hence we don't want to remove it, but escape it instead?

**Comment by nfi-opsguru** [comment]
Right, my issue is specifically about deduplicating those double slashes, but [the AWS docs](https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html) give a more thorough explanation, and it's not limited to escaping:

> Normalize URI paths according to [RFC 3986](https://tools.ietf.org/html/rfc3986). Remove redundant and relative path components.

> Is this backslash coming from AWS itself and hence we don't want to remove it, but escape it instead?

This string comes from the user's config file, not AWS.

**Comment by importhuman** [comment]
Not too experienced with AWS or Prometheus, so I looked at the [prometheus remote write docs](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_write), I'm guessing the URL that is provided is not generated via Prometheus itself, correct? So either the double slash would have to be removed in the config file itself, or we would normalize the URI as per the AWS docs. Don't think we should outright deduplicate the double slashes. 

Please correct my understanding if wrong. In the end though I think the code change would be, as suggested by Julien, escaping while doing the sigv4 sign.

**Comment by bboreham** [comment]
Closed by https://github.com/prometheus/common/pull/364
